Only check for 'other' block control slot in useHasAnyBlockControls

jeryj opened the pull request:
The useHasAnyBlockControls hook is a hefty way of checking if we are in `contentOnly` mode for the image and featured image blocks. This reduces the performance impact from https://github.com/WordPress/gutenberg/pull/58979 while we figure out what to do in a more permanent way.

Other ideas:
**1. Have a non-block toolbar popover for content-only items in image/featured image**: It _only_ uses a Replace button, and it's quite weird when Top Toolbar mode is on. For example, the Replace button is in the top toolbar here without any image block icon or other toolbar items. What am I replacing? I think this should _only_ be a floating button in a popover, not something that relies on the block toolbar. There are now [more tools being added to Pattern overrides](https://github.com/WordPress/gutenberg/pull/58998), so maybe this should be its own section within the block toolbar, or a separate block toolbar? 

<img width="988" alt="Site editor with top toolbar mode on with Replace button is in the top toolbar with no context." src="https://github.com/WordPress/gutenberg/assets/967608/2b2694d1-a991-4be7-a132-ace2c3018223">

**2. Write specific overrides into the block toolbar to account for contentOnly toolbar items** Is this a misuse of the toolbar? It seems intended to not render in contentOnly mode (as no other toolbar items show), so if we do want to use it, maybe we should specifically write in these instances rather than allow for an empty-looking toolbar?

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Improve the performance of Selecting blocks.
## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Performance improvement.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The `<MediaReplaceFlow />` used by the image and featured image are in the "other" slot. Let's not check for all slots when we can do the check once. 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

jeryj committed:
Only check for 'other' block control slot

The useHasAnyBlockControls hook is a hefty way of checking if we are in contentOnly mode for image and featured image blocks. This reduces the performance impact while we figure out what to do in a more permanent way.

jeryj commented:
@youknowriad Focus average is down 7-8ms for this PR over trunk. This can be a temporary solution that has a real improvement, but I think we need a long-term design solution for block pattern override/contentOnly toolbars.

jeryj committed:
Move block toolbar check to before rendering the block toolbar

jeryj commented on packages/block-editor/src/components/block-toolbar/index.js:
```diff
@@ -239,6 +234,11 @@ export function PrivateBlockToolbar( {
  * @param {string}  props.variant        Style variant of the toolbar, also passed to the Dropdowns rendered from Block Toolbar Buttons.
  */
 export default function BlockToolbar( { hideDragHandle, variant } ) {
+	const hasBlockToolbar = useHasBlockToolbar();
+	if ( ! hasBlockToolbar ) {
+		return null;
+	}
```
Moving this check to before the `<PrivateBlockToolbar />` avoids making an unnecessary `useSelect`, as if this is false then we know the toolbar will not render.

draganescu commented on packages/block-editor/src/components/block-controls/use-has-block-controls.js:
```diff
@@ -2,34 +2,27 @@
+ * TODO: Remove this hook, as having a toolbar with only a Replace button is a
+ * misuse of the toolbar.
+ */
 export function useHasAnyBlockControls() {
```
The name of the hook, if the change remains, should be updated to something like `useHasOtherBlockControls`.

jeryj replied:
I considered it too. You're right, the name should get updated if we go with this change.

draganescu reviewed (commented):
Can't plugins register custom controls in any groups? Would this change not hide the block toolbar if the controls are not in `other`?

Mamaduka reviewed (changes requested):
PR assumes that content related to `BlockControls` will only be registered in the `other` group. Unfortunately, we can't ensure that. For example, Quote adds citation control in the `block` group, similar to other blocks using the `Caption` component.

jeryj commented:
@Mamaduka 

> PR assumes that content related to BlockControls will only be registered in the other group. Unfortunately, we can't ensure that. For example, Quote adds citation control in the block group, similar to other blocks using the Caption component.

It's still possible to add items to all the `BlockControls` slots. The only time this check for 'other' `BlockControls` slot is applicable is when the canvas is in `contentOnly` mode, which was assumed to mean "no editing is allowed." There are a limited amount of blocks that allow editing in this state, and all of them that I saw used the `other` slot, [like the image block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/image.js#L561-L565).

So, I believe everything is still accounted for by only checking the other slot, since we only care about this in contentOnly mode. It's already a hack of the toolbar, IMO, so I don't think we need to account for third party usage of this.

jeryj committed:
Rename to useHasOtherBlockControls

draganescu commented:
I share @Mamaduka 's concern in my comment, that we can't assume that no controls in `other` means no controls at all, even if this is accidentally true today for core blocks.

jeryj commented:
I'll rebase and see. Maybe I was looking at the old block code.

> we can't assume that no controls in other means no controls at all, even if this is accidentally true today for core blocks.

My point is that it shouldn't have been allowed to be done this way in the first place, so let's prevent future contentOnly mode blocks from using other slots to contain the issue while we come up with a better solution.

jeryj commented:
@Mamaduka - I've rebased. Could you direct me to the changes to the Quote block you are concerned about related to this PR? I've tested it out and don't see anything different between this PR and trunk, and [these are the only `<BlockControls />` I see for the Quote block](https://github.com/WordPress/gutenberg/blob/perf/use-has-any-block-controls/packages/block-library/src/quote/edit.js#L97-L104).

Mamaduka commented:
> My point is that it shouldn't have been allowed to be done this way in the first place, so let's prevent future contentOnly mode blocks from using other slots to contain the issue while we come up with a better solution.

The `BlockControls` pre-date the `contentOnly` editing mode. Enforcing the limitation now will be a breaking change.

Here's the example block for testing.

```html
<!-- wp:group {"templateLock":"contentOnly","layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:paragraph -->
<p>Hey there!</p>
<!-- /wp:paragraph --></blockquote>
<!-- /wp:quote --></div>
<!-- /wp:group -->
```

The block toolbar is visible on the trunk with the Quote block's content controls.

![CleanShot 2024-04-04 at 16 12 40](https://github.com/WordPress/gutenberg/assets/240569/13dafcc6-1056-4728-9c38-8f8bd586da32)

The toolbar is no longer visible on this branch.

![CleanShot 2024-04-04 at 16 14 29](https://github.com/WordPress/gutenberg/assets/240569/2539bc7d-1194-4c21-ba8d-21d8b18b496a)

jeryj commented:
Thanks for helping me understand this, @Mamaduka! Sorry if I'm coming across as combative. I really do appreciate your help and want to figure this out!

> The BlockControls pre-date the contentOnly editing mode. Enforcing the limitation now will be a breaking change.

Even though `BlockControls` pre-dates `contentOnly`, `contentOnly` is _supposed_ to hide all block tools:

> 'contentOnly': Hides all non-content UI, e.g. auxiliary controls in the toolbar, the block movers, block settings.
> https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/README.md#L893-L894

> contentOnly — prevents all operations. Additionally, the block types that don't have content are hidden from the list view and can't gain focus within the block list. Unlike the other lock types, this is not overrideable by children.
> https://github.com/WordPress/gutenberg/blob/perf/use-has-any-block-controls/packages/block-editor/src/components/inner-blocks/README.md#L128

My argument is that  you _can_ use the block toolbar in contentOnly mode, but you're not _supposed_ to be able to. The quote block having a block toolbar when in `contentOnly` mode on trunk seems like a bug from my reading of contentOnly mode.

jeryj commented:
This has been really helpful for me to better understand the intention of contentOnly. So, any tools related to editing _content_ are allowed, but not layout. And the block toolbar doesn't know which tools should be allowed in contentOnly mode or not, so any slots need to be available for rendering. Does this sound right?

If that's the case, then would a way to indicate a tool is a content tool vs layout tool be useful? Are there any other ways we can determine this without checking all the slots?

Mamaduka commented:
@jeryj, no worries at all. I appreciate our discussion here.

> This has been really helpful for me to better understand the intention of contentOnly. So, any tools related to editing content are allowed, but not layout. And the block toolbar doesn't know which tools should be allowed in contentOnly mode or not, so any slots need to be available for rendering.

That sounds correct.

Valtio powers the SlotFills store; I wonder if we can compute a derived state that indicates the existence of Fills - `useHasSlotFill( slotName )`. However, it might be problematic since each `BlockControl` group is its own slot. cc @youknowriad, @jsnajdr

jeryj committed:
Try removing checking for slot fills on contentOnly

What breaks? Can we solve that in a different way without checking for the presence of fills in any slots?

jeryj commented:
I'm going to try removing the slot fills check entirely and see if we can fix its issues a different way.

jeryj commented:
Closing in favor of https://github.com/WordPress/gutenberg/pull/60717
